File: LineAutobotEmbeddings/gcp-cloud-function/js/openAiService.js
const { OpenAI } = require('openai');

class OpenAiService {
  constructor() {
    this.openai = new OpenAI(); // OpenAIのAPIキーを使用してインスタンスを初期化
  }

  /**
   * QAデータ（質問と回答のペア）を埋め込み（ベクトル形式）に変換するメソッド
   * @param {Array} qaData - 質問と回答のペアのリスト。各要素は { question, answer } のオブジェクト。
   * @returns {Promise<Array>} 埋め込みデータを含む配列。各質問が埋め込みベクトルとともに保存される。
   */
  async generateEmbeddings(qaData) {
    const embeddings = [];

    for (const { question, answer } of qaData) {
      const response = await this.openai.embeddings.create({
        model: "text-embedding-ada-002",
        input: question,
      });

      embeddings.push({
        question,
        answer,
        embedding: response.data[0].embedding,
      });
    }

    return embeddings;
  }
}

module.exports = OpenAiService;

File: src/functions/embeddings/services/openai_service.py
from openai import OpenAI

class OpenAiService:
    def __init__(self):
        self.openai = OpenAI()  # OpenAIのAPIキーを使用してインスタンスを初期化

    async def generate_qa_embeddings(self, qa_data):
        """
        QAデータ（質問と回答のペア）を埋め込み（ベクトル形式）に変換するメソッド
        :param qa_data: 質問と回答のペアのリスト。各要素は { 'question': str, 'answer': str } の辞書。
        :return: 埋め込みデータを含むリスト。各質問が埋め込みベクトルとともに保存される。
        """
        embeddings = []

        for qa_pair in qa_data:
            question_response = await self.generate_embeddings(self, qa_pair["question"])
            answer_response = await self.generate_embeddings(self, qa_pair["answer"])
            embeddings.append({
                'question': qa_pair['question'],
                'answer': qa_pair['answer'],
                'question_embedding': question_response,
                'answer_embedding': answer_response
            })

        return embeddings

    async def generate_embeddings(self, text):
        response = await self.openai.embeddings.create(
            model="text-embedding-ada-002",
            input=text
        )
        return response.data[0].embedding